bioblend/galaxy/jobs/__init__.py

```diff
@@ -70,6 +70,70 @@ def show_job(self, job_id, full_details=False):
 
         return self._get(id=job_id, params=params)
 
+    def _build_for_rerun(self, job_id):
+        """
+        Get details of a given job that can be used to rerun the corresponding tool.
+
+        :type job_id: str
+        :param job_id: job ID
+
+        :rtype: dict
+        :return: A description of the given job, with all parameters required to rerun.
+
+        """
+        url = '/'.join((self._make_url(job_id), 'build_for_rerun'))
+        return self._get(url=url)
+
+    def rerun_job(self, job_id, remap=False, tool_inputs_update=None, history_id=None):
+        """
+        Rerun a job.
+
+        :type job_id: str
+        :param job_id: job ID
+
+        :type remap: bool
+        :param remap: when ``True``, the job output(s) will be remapped onto the dataset(s)
+          created by the original job; if other jobs were waiting for this job to finish
+          successfully, they will be resumed using the new outputs of this tool run. When
+          ``False``, new job output(s) will be created. Note that if Galaxy does not permit
+          remapping for the job in question, specifying ``True`` will result in an error.
+
+        :type tool_inputs_update: dict
+        :param tool_inputs_update: dictionary specifying any changes which should be
+          made to tool parameters for the rerun job.
+
+        :type history_id: str
+        :param history_id: ID of the history in which the job should be executed; if
+          not specified, the same history will be used as the original job run.
+
+        :rtype: dict
+        :return: Information about outputs and the rerun job
+        .. note::
+          This method can only be used with Galaxy ``release_20.09`` or later.
+        """
+        job_rerun_params = self._build_for_rerun(job_id)
+        job_inputs = job_rerun_params['state_inputs']
+
+        if remap:
+            if not job_inputs['job_remap']:
+                raise ValueError('remap was set to True, but this job is not remappable.')
+            job_inputs['rerun_remap_job_id'] = job_id
+
+        if tool_inputs_update:
+            for input_param, input_value in tool_inputs_update.items():
+                job_inputs[input_param] = input_value
+        url = '/'.join((self.gi.url, 'tools'))
+        payload = {
+            "history_id": history_id if history_id else job_rerun_params['history_id'],
+            "tool_id": job_rerun_params['id'],
+            "inputs": job_inputs
+
+        }
+        params = {
+            "input_format": '21.01'
+        }
```

Comment:
I think you need this here (instead of inside the `payload` dict) because of https://github.com/galaxyproject/galaxy/pull/10720/files#diff-f12168394d9b932c4e06147dc6b4fda47b6947db99d7fab633789d62815bd009R539 which should have been `input_format = payload.get('input_format', 'legacy')`

Can we fix Galaxy 21.01 instead?

Reply:
Indeed - are there some guidelines for which information should be conveyed via `params` and which via `payload`?

Reply:
I created a PR here: https://github.com/galaxyproject/galaxy/pull/11332

Reply:
GET requests use `params`, POST and PUT requests use `payload`

Reply:
That makes sense to me, but then why do many of the methods defining POST API endpoints (e.g. that `create` one) take `**kwds` arguments?

Reply:
It may be due to how routing is implemented, but I'm not very sure to be honest.